### ISSUE TITLE
Loss=nan

### ISSUE THREAD

**Issue by milkyash628** [valeoai/WaffleIron#18]
When I try to train  waffleiron12-256 model on SemanticKITTI dataset, if I set batchsize=4, the loss for the first few epochs validation periods is Nan, but the epochs will return to normal after that. When I set batchsize=2, the training will continue normally without loss=Nan. May I ask how to solve this problem?

**Comment by gpuy** [comment]
Hi @milkyash628 

I will need a few more details.

- Is it the training loss or the validation loss? 
- Are the mIoU curves ok?
- Did it happen just once? Is the issue reproducible on your side?
- Were there changes done in the code?
- If the problem is reproducible when using the released code as is, can you share your config file for the architecture, the version of pytorch and the command line you used to launch the training? I will try to reproduce it.

On my side, I have not experienced NaNs with WaffleIron with a feature size of <= 384. However, I had sometimes encountered NaNs with the much larger backbone WI-48-768. For WI-48-768, changing the batchnorms to layernorms in the core of the backbone stabilized the training. I suspect the reason was numerical instabilities when training WI-48-768 with batchnorms in mixed precision.

On your side:

- If the GPU memory allows it, you can try a training without mixed precision. It suffices to remove the argument `--fp16` when calling `launch_train.py`.
- You can also try using layernorms instead of batchnorms in the core of the backbone. It suffices to add a line `layer_norm=True` when creating the [backbone](https://github.com/valeoai/WaffleIron/blob/master/launch_train.py#L176).
- If the problem appears only on the training loss while having mIoU (train/val) and validation loss curves with a normal behaviour, then it might just be a few samples that create numerical instabilities when computing the loss. To identify more easily if the problem concerns only a few samples during training, you can apply the change below to better monitor the loss.
 
---
 
*New monitoring of the loss:*

In this released code, the [monitored loss](https://github.com/valeoai/WaffleIron/blob/master/utils/trainer.py#L178) is accumulated over the whole epoch: one NaN will affect the monitored loss for the whole epoch. You can get a more finegrained monitoring by resetting `running_loss = 0.0` [here](https://github.com/valeoai/WaffleIron/blob/master/utils/trainer.py#L196) and replacing `(it + 1)` by `print_freq` [here](https://github.com/valeoai/WaffleIron/blob/master/utils/trainer.py#L202). After this change, if the training loss decreases correctly with the exception of just a few NaNs then this will confirm that just a few samples are problematic.

**Comment by milkyash628** [comment]
Hi
Here are some additional details：

1. It's the validation loss;

2. The mIoU curve is normal. The validation loss curve has problems

3. I tried to reproduce waffleiron 24-256 and 12-256, as well as single gpu and multiple gpus, but when I set bs=4, the validation loss for previous cycles was Nan;

4. Because of the third reason, I now change the code based on bs=2, as long as I change bs=4, Nan will be generated

5. Due to the problem that occurred during reproduction, I only changed the configuration file depth: 12, instance_cutmix: False, and the version of pytorch to 2.2.2,

Thank you for your reply. I will make improvements according to your suggestions.

**Comment by milkyash628** [comment]
I have tried to use Layernorms in the core of backbone and have run a few epochs so far, the loss is normal. The problem seems to be solved.

**Comment by gpuy** [comment]
Thank you for the additional details.

I will relaunch a training using your setting and batchnorms to see if I can reproduce the issue.

Good to read that layernorms are solving the problem for now. 

About layernorms, I forgot to mention that, after the training finishes, if you need to re-evaluate the final checkpoint with the script `launch_train.py` and the arguments ` --restart --eval`, you will need to comment the 4 lines between [here](https://github.com/valeoai/WaffleIron/blob/master/launch_train.py#L247) and [here](https://github.com/valeoai/WaffleIron/blob/master/launch_train.py#L250). Actually, these 4 lines can stay commented when using batchnorms too. I will try to update the code soon to make it fully compatible with layernorms.

**Comment by milkyash628** [comment]
Thanks for your advice! Everything is normal at present, and I will continue to work according to your advice.

**Comment by gpuy** [comment]
I could investigate a bit the issue you reported when training WI-12-256 with batchnorms, a batch size of 4, on SemanticKITTI.

- with pytorch 1.11, I did not see any issues.
- with pytorch 2.2, I did obtain a NaN on the validation loss after the first epoch. The issue occurs in a specific batch. The training seems to continue without any issue though. The evaluation is back to normal after a few epochs.

Interestingly, when relaunching an evaluation with a problematic checkpoint in the second bullet point above:
- with pytorch 2.2, the problem disappears when evaluating with a batch size <=2.
- with pytorch 2.2, the problem also disappears when removing mixed precision (whatever the batch size).
- with pytorch 1.11, the evaluation goes fine whatever the batch size, with or without mixed precision.

I need more time to investigate further and understand better the origin of the numerical instability.

While waiting for a more satisfying solution:
- I believe it is fine to continue with layernorms. So far, I haven't noticed any significant difference in performance when switching from batchnorm to layernorm.
- If batchnorms are required, using pytorch 1.11 seems the best option (training is however slower). As long as the problem is restricted to the validation step in the first few epochs, using pytorch 2 should be ok too.

**Comment by milkyash628** [comment]
Thank you for investigating this issue!  I didn't expected that the pytorch version is the reason,, so according to you, it really doesn't matter if only the first few epoch validation periods go wrong?  Finally, I hope you can understand the cause of the problem.

**Comment by gpuy** [comment]
We have found the origin of the numerical instabilities. It was located in the implementation of the projection from 3D to 2D used for pytorch 2. 

For batch sizes >= 2, the point clouds are zero padded to have the same size. The added fake points are all projected in the same cells/pixels in 2D. During the accumulation step, the value could increase too much until getting `inf` in mixed precision when too many fake points exist.

The true reason is thus not about batchnorm vs layernorm nor the pytorch version. Technically, this issue could have popped up with "true" points with batchnorm or layernorm.

I just pushed a fix. I tested the new code with the public checkpoints. I also relaunched a training in the problematic setting you reported. It seems solved on my side. Let me know if you notice a problem. If so, we might have to force the execution of this function in float32.

---

Along with this fix, there is now the possibility to define the type of normalization in the backbone directly in the config files. For example:
```
waffleiron:
  which_norm: "layernorm"
  nb_channels: 256
  depth: 12
```
`which_norm` can also be set to `batchnorm`. If `which_norm` is not defined in the config file, then batchnorms are used by default.
